### ISSUE TITLE
Creating empty SparseAxisArray not possible

### ISSUE THREAD

**Issue by junglegobs** [JuliaOpt/JuMP.jl#2169]
I know this is a weird thing to complain about, but before JuMP 20.1 I used to be able to do this:
```julia
m = Model()
@variable(m, v[i=[], j=[]; j <= 2])
```
Now I get the following error:
```julia
ERROR: MethodError: no method matching JuMP.Containers.SparseAxisArray(::Dict{Any,Any})
Closest candidates are:
  JuMP.Containers.SparseAxisArray(::Dict{K<:Tuple{Vararg{Any,N}},T}) where {T, N, K<:Tuple{Vararg{Any,N}}} at /home/u0128861/.julia/packages/JuMP/CZ8vV/src/Containers/SparseAxisArray.jl:38
Stacktrace:
 [1] container(::Function, ::JuMP.Containers.NestedIterator{Tuple{getfield(Main, Symbol("##282#286")),getfield(Main, Symbol("##283#287"))},getfield(Main, Symbol("##284#288"))}, ::Type{JuMP.Containers.SparseAxisArray}) at /home/u0128861/.julia/packages/JuMP/CZ8vV/src/Containers/container.jl:94
 [2] container(::Function, ::JuMP.Containers.NestedIterator{Tuple{getfield(Main, Symbol("##282#286")),getfield(Main, Symbol("##283#287"))},getfield(Main, Symbol("##284#288"))}) at /home/u0128861/.julia/packages/JuMP/CZ8vV/src/Containers/container.jl:65
 [3] top-level scope at /home/u0128861/.julia/packages/JuMP/CZ8vV/src/macros.jl:79
```

This is obviously a trivial / silly complaint - the only reason I mention it is because it's useful for me to be able to make empty variables and expressions. I have a very general form of an energy system optimisation problem, and if for example some units (e.g. storage) don't exist then the form of my equations was still valid given JuMP's previous behaviour. I could of course just adapt my code to do something more sensible if it realises it's trying to make an empty sparse variable (e.g. return an array of zeros).

**Comment by odow** [comment]
How are you using `v` if it has not terms?

This is hard, because we don't know what the keys are.

One work-around if for you to go
```Julia
v = Dict((i,j) => @variable(m) for i=[], j=[] if j <= 2)
```

Alternatively (I'm not convinced that this is a good idea to add), we could go:
```Julia
function Containers.SparseAxisArray(d::Dict{Any,Any})
    return Containers.SparseAxisArray(Dict{Tuple{},Any}(d))
end
```

**Comment by junglegobs** [comment]
`v` still appears in an energy balance, but it's just zero. I eventually went for a (slightly) more transparent workaround where if the set is empty it just returns a custom `EmptyContainer` type. This returns `zero(T::Type)` no matter what you index it with.
```
"""
        mutable struct EmptyContainer

Returns zero(T) or nothing if T <: Nothing.
"""
mutable struct EmptyContainer

    returnType::Type

    """
        EmptyContainer(T::Type)

    Create an EmptyContainer.
    """
    function EmptyContainer(T::Type)
        ec = new()
        if (T <: Nothing) == false
            try
                zero(T)
            catch
                error("Cannot return zero of type $T")
            end
        end
        ec.returnType = T
        return ec
    end

    function EmptyContainer()
        EmptyContainer(Nothing)
    end

    function Base.getindex(ec::EmptyContainer, args...)
        if ec.returnType <: Nothing
            return nothing
        else
            return zero(ec.returnType)
        end
    end

    function Base.print(io::IO, ec::EmptyContainer)
        print(io, "An EmptyContainer")
    end

    function Base.show(io::IO, ec::EmptyContainer)
        print(io, ec)
    end

    function Base.length(ec::EmptyContainer)
        return 0
    end
end
```

**Comment by mlubin** [comment]
This is an unintended regression from 0.20. Should be fixed for 0.21.x.

**Comment by blegat** [comment]
Related bug:
```julia
julia> @variable(m, [i=1:0, j=[]], container = SparseAxisArray)
ERROR: MethodError: no method matching JuMP.Containers.SparseAxisArray(::Dict{Any,Any})
Closest candidates are:
  JuMP.Containers.SparseAxisArray(::Dict{K,T}) where {T, N, K<:Tuple{Vararg{Any,N}}} at /home/blegat/.julia/packages/JuMP/CZ8vV/src/Containers/SparseAxisArray.jl:38
Stacktrace:
 [1] container(::Function, ::JuMP.Containers.VectorizedProductIterator{Tuple{Base.OneTo{Int64},Array{Any,1}}}, ::Type{JuMP.Containers.SparseAxisArray}) at /home/blegat/.julia/packages/JuMP/CZ8vV/src/Containers/container.jl:94
 [2] top-level scope at /home/blegat/.julia/packages/JuMP/CZ8vV/src/macros.jl:45
```
